### ISSUE TITLE
Disallow mutating state of Blob/File objects

### ISSUE THREAD

**Issue by annevk** [w3c/FileAPI#75]
In particular, I think we should no longer add something like `close()` and make that an invariant of these objects.

That way there's no longer a need for defensive cloning (through serializing/deserializing) of these objects, which helps with https://github.com/whatwg/url/issues/277.

Making a note of this somewhere in the document seems sufficient.

**Comment by mkruisselbrink** [comment]
I agree that defining Blob/File as fully immutable (which they more or less already are currently) would definitely be a nice thing. I'm not sure if we can really get rid of defensive cloning though. For example reading a File can always cause a NotReadableError if the snapshot state of the File no longer matches the state of the underlying storage (although the whole concept of this snapshot state isn't defined very well. I don't think we want to start requiring user agents to actually do defensive cloning themselves every time they create a File).

**Comment by annevk** [comment]
Defensive cloning doesn't help with snapshot state I think, since we wouldn't want to require cloning to copy the bytes, cloning should just copy a pointer.

**Comment by mkruisselbrink** [comment]
Ah yes, good point. So yeah, I should definitely add a note somewhere to make it explicit that blob's are immutable, noting that the snapshot state might still cause the same operation on the same blob to behave differently at different times.

**Comment by guest271314** [comment]
@annevk Related https://github.com/w3c/FileAPI/issues/99. I.e.g.; at Chromium/Chrome 

```
<!DOCTYPE html>
<html>
<head>
  <style>
    #info {
      font-family: Courier;
      font-size: 0.9em;
    }
  </style>
  <script>
    window.addEventListener('load', function() {
      let size = 0;
      let readFile = file => {
        const reader = new FileReader;
        reader.onload = () => alert(reader.result);
        reader.readAsText(file);
      }
      window.setInterval(function() {
        var logFile = document.querySelector('#file').files[0];
        if (logFile) {
          if (!size) {
            size = logFile.size;
          }
          if (size && logFile.size !== size) {
            size = 0;
            readFile(logFile);
          }
          document.querySelector('#info').innerHTML = '<br/>' +
            (new Date()).toString() + '<br/>Last modified: ' +
            logFile.lastModified +
            '<br/>Size: ' +
            logFile.size;
        }
      }, 1000);
    });
  </script>
</head>
<body>
  <input type="file" id="file" />
  <p id="info"></p>
</body>
</html>
```
Awaiting verification that the behaviour at Chromium/Chrome is in fact a bug, will then file bug report.

**Comment by andrew-aladev** [comment]
> I agree that defining Blob/File as fully immutable (which they more or less already are currently) would definitely be a nice thing.

I don't think so. You like functional programming. You just forgot that you are making API for whole Earth. Some people don't like it. Why do you think that you can force your programming style?

**Comment by andrew-aladev** [comment]
> In particular, I think we should no longer add something like `close()` and make that an invariant of these objects.

If you don't want to maintain blob/file state inside - people will have to invent their own bicycle wrapper that does this (with some method like `toNative()`). If you are going to make user applications more error prone - you are doing fine job.

**Comment by guest271314** [comment]
@andrew-aladev If `toNative()` only sets the `name` property of the `File` object `FormData.prototype.set` can be utilized https://github.com/w3c/FileAPI/issues/126#issuecomment-480956457
